text/0000-trait-alias.md

```diff
@@ -0,0 +1,266 @@
+[tokio case]([tokio example](#second-example-ergonomic-collections-and-scrapping-boilerplate):
+
+```rust
+pub trait Service {
+  type Request;
+  type Response;
+  type Error;
+  type Future: Future<Item=Self::Response, Error=Self::Error>;
+  fn call(&self, req: Self::Request) -> Self::Future;
+}
+
+trait HttpService = Service<Request = http::Request, Response = http::Response, Error = http::Error>;
+
+trait MyHttpService = HttpService<Future = MyFuture>; // assume MyFuture exists and fulfills the rules to be used in here
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Adds another construct to the language.
+
+- The syntax `trait TraitAlias = Trait` requires lookahead in the parser to disambiguate a trait from a trait alias.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- It’s possible to create a new trait that derives the trait to alias, and provide a universal `impl`:
+    
+    ```rust
+    trait Foo {}
+    
+    trait FooFakeAlias: Foo {}
+    
+    impl<T> Foo for T where T: FooFakeAlias {}
+    ```
+    
+    This works for trait objects and trait bounds only. You cannot implement `FooFakeAlias` directly
+    because you need to implement `Foo` first – hence, you don’t really need `FooFakeAlias` if you can
+    implement `Foo`.
+    
+    There’s currently no alternative to the impl problem described here.
+
+- Similar to Haskell's ContraintKinds, we could declare an entire predicate as a reified list of constraints, instead of creating an alias for a set of supertraits and predicates. Syntax would be something like `constraint Foo<T> = T: Bar, Vec<T>: Baz;`, used as `fn quux<T>(...) where Foo<T> { ... }` (i.e. direct substitution). Trait object usage is unclear.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Should we use `type` as the keyword instead of `trait`?
+
+    `type Foo = Bar;` already creates an alias `Foo` that can be used as a trait object.
+    
+    If we used `type` for the keyword, this would imply that `Foo` could also be used as a bound as well. If we use `trait` as proposed in the body of the RFC, then `type Foo = Bar;` and `trait Foo = Bar;` _both_ create an alias for the object type, but only the latter creates an alias that can be used as a bound, which is a confusing bit of redundancy.
+    
+    However, this mixes the concepts of types and traits, which are different, and allows nonsense like `type Foo = Rc<i32> + f32;` to parse.
+    
+- Which bounds need to be repeated when using a trait alias?
+
+    [RFC 1927](https://github.com/rust-lang/rfcs/pull/1927) intends to change the rules here for traits, and we likely want to have the rules for trait aliases be the same to avoid confusion.
+    
+    The `contraint` alternative sidesteps this issue.
+
+- What about bounds on type variable declaration in the trait alias? Consider the following:
+
+    ```rust
+    trait Foo<T: Bar> = PartialEq<T>;
```

Comment:
This is related to `type Foo<T: Bar>` -- and our current handling there is [not very good](https://github.com/rust-lang/rust/issues/21903). I think the fix I was hoping to emply in #21903 (but I have to work out how it would work...) is to avoid issuing a warning for things that are required for the content of the trait-alias to be well-formed (in particular, bounds you need to accommodate some project). 

In other words, if we applied the same intuition here, but declared that unused bounds are illegal (in types we may have to permit them because back compat):

```
trait Foo<T: Iterator> = PartialEq<T::Item>; // legal, `T: Iterator` implied by `T::Item`
trait Foo<T: Iterator> = PartialEq<T>; // illegal, adds `T: Iterator` from nowhere
```